UnityProject/Assets/Scripts/Messages/Server/TabUpdateMessage.cs

```diff
@@ -55,68 +108,178 @@ public static void SendToPeepers(GameObject provider, NetTabType type, TabAction
 		ElementValue[] values = null)
 	{
 
-		if (values != null)
+		switch (tabAction)
+		{
+			case TabAction.Open:
+				NetworkTabManager.Instance.Add(provider, type, recipient);
+				values = NetworkTabManager.Instance.Get(provider, type).ElementValues;
+				break;
+			case TabAction.Close:
+				NetworkTabManager.Instance.Remove(provider, type, recipient);
+				break;
+			case TabAction.Update:
+				//fixme: duplication of NetTab.ValidatePeepers
+				//Not sending updates and closing tab for players that don't pass the validation anymore
+				var validate = Validations.CanApply(recipient, provider, NetworkSide.Server);
+				if (!validate)
+				{
+					Send(recipient, provider, type, TabAction.Close);
+					return null;
+				}
+				break;
+		}
+
+		// SingleMessage: 0, start/middle: 1, end: 2
+		var id = 0;
+
+		var uniqueID = Random.Range(0, 2000);
+
+		if (ServerUniqueIdCache.Contains(uniqueID))
+		{
+			uniqueID = Random.Range(2001, 20000);
+		}
+
+		ServerUniqueIdCache.Add(uniqueID);
+
+		var elementValuesLists = new Dictionary<List<ElementValue>, int>();
+
+		if (values != null && tabAction != TabAction.Close)
 		{
 			// get max possible packet size from current transform
 			int maxPacketSize = Transport.activeTransport.GetMaxPacketSize(0);
+
 			// set currentSize start value to max TCP header size (60b)
-			int currentSize = 60;
+			float currentSize = 60f;
```

Comment:
I haven't read this much but that it's a float makes me very suspicious.

Reply:
I changed it up so it cast to float in the Math.Ceiling.